File: Solutions/Employee Free Time.cpp
/*
// Definition for an Interval.
class Interval {
public:
    int start;
    int end;

    Interval() {}

    Interval(int _start, int _end) {
        start = _start;
        end = _end;
    }
};
*/

class Solution {
public:
    // comparator function (static because inside class)
    static bool compare(pair<Interval, int> i1, pair<Interval, int> i2) {    
        // condition reversed because we want a min-heap
        return i1.first.start > i2.first.start;
    }
    
    // merge k sorted lists of intervals in schedule
    vector<Interval> mergeKSortedLists(vector<vector<Interval>> schedule) {
        int k = schedule.size();
        vector<Interval> merged;
        
        // indices into each list
        vector<int> indices(k, 0);
        
        // insert all starting elements of each list into priority queue
        // {interval, list number}
        priority_queue<pair<Interval, int>, vector<pair<Interval, int>>, decltype(&Solution::compare)> pq(compare);
        for (int i=0; i<k; i++)
            pq.push({schedule[i][0], i});
        
        // merge k sorted schedules
        while (!pq.empty()) {
            auto [interval, list_num] = pq.top();
            pq.pop();
            
            merged.push_back(interval);
            indices[list_num]++;
            if (indices[list_num] < schedule[list_num].size())
                pq.push({schedule[list_num][indices[list_num]], list_num});
        }
        
        return merged;
    }
    
    // merge intervals
    vector<Interval> mergeIntervals(vector<Interval> intervals) {        
        vector<Interval> out;
        Interval merged(intervals[0].start, intervals[0].end);
        
        for (auto &interval : intervals) {
            // overlap occurs
            if (interval.start <= merged.end)
                merged.end = max(merged.end, interval.end);
            else {
                out.push_back(merged);
                merged = interval;
            }
        }
        
        out.push_back(merged);
        return out;
    }
    
    vector<Interval> employeeFreeTime(vector<vector<Interval>> schedule) {        
        vector<Interval> intervals = mergeIntervals(mergeKSortedLists(schedule));
        
        int n = intervals.size();
        vector<Interval> out;

        // find all gaps between sorted intervals
        for (int i=0; i<n-1; i++) {
            Interval interval(intervals[i].end, intervals[i+1].start);
            out.push_back(interval);
        }
        
        return out;
    }
};

/*
we are given k sorted lists of Intervals (one for each employee).

- merge k sorted lists (such that the merged list is also sorted).
    time complexity: O(nlogk)
    
- merge intervals in this sorted list.
    time complexity: O(nk)

- find all gaps between the intervals after merging.
    time complexity: O(nk)

- the list of these gaps is the required answer.

time complexity: O(nk)
space complexity: O(nk)

https://leetcode.com/problems/employee-free-time/
*/

File: Solutions/Blind75/hard/Employee Free Time.py
"""
# Definition for an Interval.
class Interval:
    def __init__(self, start: int = None, end: int = None):
        self.start = start
        self.end = end
"""

class Solution:    
    def mergeKSorted(self, intervals):
        k = len(intervals)
        pointers = [0] * k
        
        # push into priority queue
        pq = []
        for i in range(k):
            heapq.heappush(pq, (intervals[i][0], i))
        
        merged = []
        while pq:
            interval, list_num = heapq.heappop(pq)
            merged.append(interval)
            
            # move pointer forward
            pointers[list_num] += 1
            if pointers[list_num] < len(intervals[list_num]):
                heapq.heappush(pq, (intervals[list_num][pointers[list_num]], list_num))
        
        return merged
    
    def employeeFreeTime(self, schedule: '[[Interval]]') -> '[Interval]':
        schedule = [[[interval.start, interval.end] for interval in intervals] for intervals in schedule]
        intervals = self.mergeKSorted(schedule)
        
        # merge intervals
        out = []
        merged = intervals[0]
        
        for interval in intervals:
            # overlap occurs
            if interval[0] <= merged[1]:
                merged[1] = max(merged[1], interval[1])
            else:
                out.append(Interval(merged[1], interval[0]))
                merged = interval
        
        return out
        
"""
merge sorted lists of intervals for each employee
merge intervals of this new list

k = number of employees
n = total number of intervals

time complexity: O(nlogk + n)
space complexity: O(n)
"""